casper/src/main/resources/TreeHashMap.rho

```diff
@@ -0,0 +1,220 @@
+
+new TreeHashMap, rs(`rho:registry:insertSigned:secp256k1`), uriOut in {
+  new MakeNode, ByteArrayToNybbleList, TreeHashMapSetter, TreeHashMapGetter, 
+      powersCh, storeToken, nodeGet, nodeSetBit, nodeSetMap in {
+    powersCh!([1,2,4,8,16,32,64,128,256,512,1024,2048,4096,8192,16384,32768,65536]) |
+    for (@powers <- powersCh) {
+      contract MakeNode(@initVal, @node) = {
+        @[node, *storeToken]!(initVal)
+      } |
+        
+      contract nodeGet(@node, ret) = {
+        // TODO: convert to peek
+        for (@val <- @[node, *storeToken]) {
+          @[node, *storeToken]!(val) |
+          ret!(val)
+        }
+      } |
+
+      contract nodeSetBit(@node, @bit, ack) = {
+        for (@val <- @[node, *storeToken]) {
+          // val | (1 << bit)
+          // Bitwise operators would be really nice to have!
+          @[node, *storeToken]!((val % powers.nth(bit)) +
+            (val / powers.nth(bit + 1)) * powers.nth(bit + 1) +
+            powers.nth(bit)) |
+          ack!(Nil)
+        }
+      } |
+      
+      contract nodeSetMap(@node, @newKey, @newVal, ack) = {
+        for (@val <- @[node, *storeToken]) {
+          ack!(Nil) |
+          if (val == 0) { @[node, *storeToken]!({newKey: newVal}) }
+          else { @[node, *storeToken]!(val.set(newKey, newVal)) }
+        }
+      } |
+
+      contract ByteArrayToNybbleList(@ba, @n, @len, @acc, ret) = {
+        if (n == len) {
+          ret!(acc)
+        } else {
+          ByteArrayToNybbleList!(ba, n+1, len, acc ++ [ ba.nth(n) % 16, ba.nth(n) / 16 ], *ret)
+        }
+      } |
+
+      contract TreeHashMap(@"init", @depth, ret) = {
+        new map in {
+          MakeNode!(0, (*map, [])) |
+          @(*map, "depth")!!(depth) |
+          ret!(*map)
+        }
+      } |
+
+      contract TreeHashMapGetter(@map, @nybList, @n, @len, @suffix, ret) = {
+        // Look up the value of the node at [map, nybList.slice(0, n + 1)]
+        new valCh in {
+          nodeGet!((map, nybList.slice(0, n)), *valCh) |
+          for (@val <- valCh) {
+            if (n == len) {
+              ret!(val.get(suffix))
+            } else {
+              // Otherwise check if the rest of the path exists.
+              // Bit k set means node k exists.
+              // nybList.nth(n) is the node number
+              // val & powers.nth(nybList.nth(n)) is nonzero if the node exists
+              // (val/powers.nth(nybList.nth(n))) % 2 is 1 if the node exists
+              if ((val/powers.nth(nybList.nth(n))) % 2 == 0) {
+                ret!(Nil)
+              } else {
+                TreeHashMapGetter!(map, nybList, n + 1, len, suffix, *ret)
+              }
+            }
+          }
+        }
+      } |
+
+      contract TreeHashMap(@"get", @map, @key, ret) = {
+        new hashCh, nybListCh in {
+          // Hash the key to get a 256-bit array
+          @"keccak256Hash"!(key.toByteArray(), *hashCh) |
+          for (@hash <- hashCh; @depth <- @(map, "depth")) {
```

Comment:
In the sad world we live in, every join usage is conflicting between all the channels involved. We should split that into two receives

Reply:
Done

Reply:
Done.